### ISSUE TITLE
FSDP2 issue with mp_policy, checkpoint() and float input

### ISSUE THREAD

**Issue by mori360** [pytorch/pytorch#150140]
### 🐛 Describe the bug

https://github.com/pytorch/pytorch/pull/150139

This bug happens at when using FSDP2 setting mp_policy, with checkpoint(model, input) and input to be float input.

Here are the finding of root cause

FSDP1:

- _root_pre_forward
    - _cast_forward_inputs(state.mixed_precision.param_dtype)
- _pre_forward
    - return if training state is HandleTrainingState.BACKWARD_PRE # return in the second time

FSDP2:

- _pre_forward
    - return if training state is TrainingState.PRE_BACKWARD # return in the second time
    - _root_pre_forward
    - cast_fn(self._mp_policy.param_dtype)

In FSDP2, we have the training state fist before casting mixed_precision, thus when using `checkpoint(model, input)`, during `loss.backward()`, forward would be called twice and the 2nd time won't have the cast_fn, causing dtype error.

The cast_fn uses `_cast_fp_tensor` to cast, so if the input is not in float, it won't be cast in the 1st time, there's no issue if both 1st and 2nd time are not cast

### Versions

PyTorch version: 2.8.0a0+git49d7d66
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A
